Refactor use of findGetMetricDataByID

cristiangreco opened the pull request.

thepalbi commented on pkg/job/discovery.go:
```diff
@@ -94,41 +94,63 @@ func scrapeDiscoveryJobUsingMetricData(
+	// update getMetricDatas with fetched values and timestamps
+	for _, data := range getMetricDataOutput {
+		for _, metricDataResult := range data.MetricDataResults {
+			idx := findGetMetricDataByID(getMetricDatas, *metricDataResult.Id)
```
what was `metricDataResult.Id` here? can we add a comment explaining that?

thepalbi commented on pkg/job/discovery.go:
```diff
@@ -94,41 +94,63 @@ func scrapeDiscoveryJobUsingMetricData(
+	}
+
+	// remove unprocessed/unknown elements in place (if any)
+	getMetricDatas = compactSlice(getMetricDatas)
```
the compaction is worth because `getMetricDatas` might have a really high cardinality for example in EC2 with lots of instances + we release early memory right?

cristiangreco commented:
@thepalbi I've refactored the slice compaction into a separate method (think can be used in other places) and added comments. Can you please have another pass?

thepalbi reviewed (approved):
just perfect!

cristiangreco commented on pkg/job/compact.go:
```diff
@@ -0,0 +1,21 @@
+package job
+
+// compact iterates over a slice of pointers and deletes
+// unwanted elements as per the keep function return value.
+// The slice is modified in-place without copying elements.
+func compact[T any](input []*T, keep func(el *T) bool) []*T {
+	// move all elements that must be kept at the beginning
+	i := 0
+	for _, d := range input {
+		if keep(d) {
+			input[i] = d
+			i++
+		}
+	}
+	// nil out any left element
+	for j := i; j < len(input); j++ {
+		input[j] = nil
+	}
+	// set new slice length to allow released elements to be collected
+	return input[:i]
+}
```
Did a benchmark before merging the PR and this showed zero allocations. Anyway I'm not entirely satisfied with the memory usage show here -- as I'm touching this part of the code in other PRs I might edit or just revert this implementation.
